cli/command/trust/key_generate.go

```
@@ -0,0 +1,130 @@
+		},
+	}
+	flags := cmd.Flags()
+	flags.StringVarP(&options.directory, "dir", "d", "", "Directory to generate key in, defaults to current directory")
+	return cmd
+}
+
+// key names can use alphanumeric + _ + - characters
+var validKeyName = regexp.MustCompile(`^[a-zA-Z0-9\_]+[a-zA-Z0-9\_\-]*$`).MatchString
+
+// validate that all of the key names are unique and are alphanumeric + _ + -
+// and that we do not already have public key files in the current dir on disk
+func validateKeyArgs(keyName string, cwdPath string) error {
+	if !validKeyName(keyName) {
+		return fmt.Errorf("key name \"%s\" must not contain special characters", keyName)
+	}
+
+	pubKeyFileName := keyName + ".pub"
+	if _, err := os.Stat(filepath.Join(cwdPath, pubKeyFileName)); err == nil {
+		return fmt.Errorf("public key file already exists: \"%s\"", pubKeyFileName)
+	}
+	return nil
+}
+
+func setupPassphraseAndGenerateKeys(streams command.Streams, opts keyGenerateOptions) error {
+	targetDir := opts.directory
+	// if the target dir is empty, default to CWD
+	if targetDir == "" {
+		cwd, err := os.Getwd()
+		if err != nil {
+			return err
+		}
+		targetDir = cwd
+	}
+	return validateAndGenerateKey(streams, opts.name, targetDir)
+}
+
+func validateAndGenerateKey(streams command.Streams, keyName string, workingDir string) error {
+	freshPassRetGetter := func() notary.PassRetriever { return trust.GetPassphraseRetriever(streams.In(), streams.Out()) }
+	if err := validateKeyArgs(keyName, workingDir); err != nil {
+		return err
+	}
+	fmt.Fprintf(streams.Out(), "\nGenerating key for %s...\n", keyName)
+	// Automatically load the private key to local storage for use
+	privKeyFileStore, err := trustmanager.NewKeyFileStore(trust.GetTrustDirectory(), freshPassRetGetter())
+	if err != nil {
+		return err
+	}
+
+	pubPEM, err := generateKeyAndOutputPubPEM(keyName, privKeyFileStore)
+	if err != nil {
+		fmt.Fprintf(streams.Out(), err.Error())
+		return fmt.Errorf("Error generating key for: %s", keyName)
```

I don't think it's correct to print an error on stdout. This should return an error without the `fmt.Fprintf()`

```go
return errors.Wrapf(err, "failed to generate key for %s", keyName)
```